### PR TITLE
deps: upgrade grpc-ecosystem/go-grpc-middleware to v2 and migrate to grpc-ecosystem/go-grpc-middleware/providers/prometheus

### LINKED ISSUE #17883: Use of deprecated grpc-ecosystem/go-grpc-middleware v1 and grpc-ecosystem/go-grpc-prometheus
### What would you like to be added?

Upgrade grpc-ecosystem/go-grpc-middleware to v2 and migrate grpc-ecosystem/go-grpc-prometheus to grpc-ecosystem/go-grpc-middleware/providers/prometheus.

### Why is this needed?

Trying to get rid of deprecated go-grpc-middleware in https://github.com/grafana/grafana and this is the first stop on the path:

```
go mod why github.com/grpc-ecosystem/go-grpc-middleware
# github.com/grpc-ecosystem/go-grpc-middleware
github.com/grafana/grafana/pkg/apiserver/storage/file
k8s.io/apiserver/pkg/storage/storagebackend
k8s.io/apiserver/pkg/storage/etcd3
k8s.io/apiserver/pkg/storage/etcd3.test
go.etcd.io/etcd/server/v3/embed
go.etcd.io/etcd/server/v3/etcdserver/api/v3rpc
github.com/grpc-ecosystem/go-grpc-middleware
```

### PR DESCRIPTION
**Pull request by marefr**
Upgrades github.com/grpc-ecosystem/go-grpc-middleware to v2 and migrates the deprecated github.com/grpc-ecosystem/go-grpc-prometheus to github.com/grpc-ecosystem/go-grpc-middleware/providers/prometheus.

This is my first contribution to this project and tried to follow the contribution guidelines as good as I could. Please let me know if I missed something and/or suggested changes.

Closes #17883

### CODE REVIEW

File: server/etcdmain/grpc_proxy.go [hunk 1]
```diff
@@ -461,32 +460,32 @@ func newGRPCProxyServer(lg *zap.Logger, client *clientv3.Client) *grpc.Server {
...
+
 	if grpcProxyEnableLogging {
+		opts := []grpc_logging.Option{
+			grpc_logging.WithLogOnEvents(grpc_logging.StartCall, grpc_logging.FinishCall),
+		}
+
 		grpcChainStreamList = append(grpcChainStreamList,
-			grpc_ctxtags.StreamServerInterceptor(),
-			grpc_zap.PayloadStreamServerInterceptor(lg, alwaysLoggingDeciderServer),
+			grpc_logging.StreamServerInterceptor(interceptorLogger(lg), opts...),
```
Comment (marefr) [inline server/etcdmain/grpc_proxy.go:479]: Looks like the old `grpc_ctxtags.StreamServerInterceptor` added a field for `peer.address`, https://github.com/grpc-ecosystem/go-grpc-middleware/blob/d42ae9d517069c2bd7f9339147a0eafa86b3d4a3/tags/interceptors.go#L70-L76.

The new logging interceptor adds this field as well, hence why I've removed the old grpc_ctxtags interceptors: 
https://github.com/grpc-ecosystem/go-grpc-middleware/blob/7da22cf3f3d3ae190467d9c7a3ea749b3d0e63b5/interceptors/logging/interceptors.go#L168-L170

File: server/embed/etcd.go [hunk 2]
```diff
@@ -818,10 +817,6 @@ func (e *Etcd) pickGRPCGatewayServeContext(splitHTTP bool) *serveCtx {
 }
 
 func (e *Etcd) serveMetrics() (err error) {
-	if e.cfg.Metrics == "extensive" {
-		grpc_prometheus.EnableHandlingTimeHistogram()
-	}
```
Comment (marefr) [inline server/embed/etcd.go:823]: This part I struggled the most with. For now I'm passing this config down to `v3rpc.Server` as a string argument which doesn't feel great.

File: server/etcdserver/api/v3rpc/grpc.go [hunk 3]
```diff
@@ -35,24 +35,34 @@ const (
...
 	opts = append(opts, grpc.CustomCodec(&codec{}))
 	if tls != nil {
 		opts = append(opts, grpc.Creds(credentials.NewTransportCredential(tls)))
 	}
+
+	promOpts := []grpc_prometheus.ServerMetricsOption{}
+
+	if metrics == "extensive" {
+		promOpts = append(promOpts, grpc_prometheus.WithServerHandlingTimeHistogram())
+	}
```
Comment (marefr) [inline server/etcdserver/api/v3rpc/grpc.go:49]: This moved from server/embed/etcd.go, see the other comment.

File: tests/framework/integration/testing.go [hunk 4]
```diff
@@ -151,3 +153,94 @@ func NewClient(t testing.TB, cfg clientv3.Config) (*clientv3.Client, error) {
 	}
 	return clientv3.New(cfg)
 }
+
+// SettableLoggerV2 is thread-safe.
+type SettableLoggerV2 interface {
```
Comment (marefr) [inline tests/framework/integration/testing.go:158]: Copied from https://github.com/grpc-ecosystem/go-grpc-middleware/tree/v1.4.0/logging/settable

File: server/etcdmain/logger_interceptor.go
Comment (serathius) [inline server/etcdmain/logger_interceptor.go:1]: Expect this code to land in some `logutil` package. We have one for client `client/pkg/logutil`. I expect it's in client due to legacy, we could move it to `pkg/logutil`

**Comment by serathius** [comment]
We need to make sure we don't drop any metrics.

For this kind of PR I expect to see:
* A list of metrics before and after the change, including names, types, labels to make sure they don't change.
* Diff between the list if there is some differences
* Best an e2e test that prevents a regression.

**Comment by marefr** [comment]
@ahrtr yes that sounds like a good plan 👍

**Comment by marefr** [comment]
I'll close this and open a new PR for upgrading grpc-ecosystem/go-grpc-middleware to v2 as a first step.

**Pull request closed without merging**